### PR TITLE
rate limit CQL extension support

### PR DESCRIPTION
**Pull request by wprzytula**
Since recently, Scylla has begun to support a new CQL error type, RateLimitError. For this error to be recognised by the driver properly, a proper negotiation is needed. It was implemented similarly to the already-present LWT optimisation extension negotiation.

### CODE REVIEW

File: scylla.go [hunk 1]
```diff
@@ -290,15 +345,15 @@ func (p *scyllaConnPicker) maybeReplaceWithLessBusyConnection(c *Conn) *Conn {
...
+	if alternative == nil || alternative.AvailableStreams()*120 > c.AvailableStreams()*100 {
 		return c
 	} else {
 		return alternative
 	}
 }
 
 func isHeavyLoaded(c *Conn) bool {
-    return c.streams.NumStreams / 2 > c.AvailableStreams();
+	return c.streams.NumStreams/2 > c.AvailableStreams()
```
Comment (piodul) [inline scylla.go:356]: It looks like some code was not formatted properly before and here you gofmt-ed it. Could you move this change to a separate commit?

File: errors.go [hunk 2]
```diff
@@ -196,3 +196,28 @@ type RequestErrCASWriteUnknown struct {
 	Received    int
 	BlockFor    int
 }
+
+type opType uint8
+
+const (
+	opTypeRead  opType = 0
+	opTypeWrite opType = 1
+)
```
Comment (piodul) [inline errors.go:205]: In Go, identifiers starting with a lowercase letters are private. I think those (opType, opTypeRead, opTypeWrite) should be public.

**Review by piodul — COMMENTED** [review]
Did you test it against a real Scylla node? If so, could you describe what you did?

File: scylla.go (hunk 1 above)
Comment (wprzytula) [reply scylla.go:356]: If I run gofmt, this is what I get:
```
modified:   batch_test.go
modified:   cass1batch_test.go
modified:   cassandra_only_test.go
modified:   connectionpool_test.go
modified:   debug_off.go
modified:   debug_on.go
modified:   errors_test.go
modified:   events_ccm_test.go
modified:   example_lwt_batch_test.go
modified:   example_nulls_test.go
modified:   example_paging_test.go
modified:   example_set_test.go
modified:   example_udt_struct_test.go
modified:   exec_test.go
modified:   fuzz.go
modified:   host_source_gen.go
modified:   host_source_test.go
modified:   integration_cassandra_only_test.go
modified:   integration_test.go
modified:   metadata_cassandra.go
modified:   metadata_cassandra_test.go
modified:   metadata_scylla.go
modified:   metadata_scylla_test.go
modified:   recreate.go
modified:   recreate_test.go
modified:   scylla_shard_aware_port_integration_test.go
modified:   scylla_shard_aware_port_mocked_test.go
modified:   session_test.go
modified:   session_unit_test.go
modified:   stress_test.go
modified:   topology_test.go
modified:   tuple_test.go
modified:   udt_test.go
modified:   uuid_test.go
modified:   wiki_test.go
```
It seems that gofmt's usage was abandoned in this repo quite a long time ago. Upstream probably lives with it, and so I'm afraid we should, too. I'll just remove that fmt fix from this PR.

**Comment by wprzytula** [comment]
> Did you test it against a real Scylla node? If so, could you describe what you did?

I've tested it against the proxy only, after I confirmed the proxy test works properly with Rust driver implementation. You can view the test here: [my branch].(https://github.com/wprzytula/scylla-rust-driver/tree/gocql-tests)

**Comment by piodul** [comment]
> > Did you test it against a real Scylla node? If so, could you describe what you did?
> 
> I've tested it against the proxy only, after I confirmed the proxy test works properly with Rust driver implementation. You can view the test here: [my branch](https://github.com/wprzytula/scylla-rust-driver/tree/gocql-tests).

I think it would be best to test it out on a real Scylla anyway. It looks like Scylla 5.1.0 was pushed to the docker hub today - this version should contain the per-partition rate limiting feature.

The easiest way to test this would be to run scylla-bench built with the modified driver (you can modify the go.mod so that it uses a locally checked out version of the driver) so that it writes/reads only to a single partition. After running it once you can then run `ALTER TABLE <scylla bench table name> WITH per_partition_rate_limit = {'max_reads_per_second': 1, 'max_writes_per_second': 1}`, then run scylla-bench again and observe errors being printed to logs.

**Comment by wprzytula** [comment]
> 

Done. Having done as you suggested, I got `Per-partition rate limit reached for read in table scylla_bench.test, rejected by coordinator`, so I guess it's working.

**Review by piodul — APPROVED** [review]

File: scylla.go (hunk 1 above)
Comment (martin-sucha) [reply scylla.go]: We could run gofmt upstream and enforce it in CI. Would probably cause one-time merge conflict though.
